File: lowestAncestorC++/lowestAncestor.cpp
//C++ program run by entering "lowestAncestor <firstNode> <secondNode> <thirdNode> ...
//It then builds binary tree based on input and asks for 2 nodes to calulate LCA of.
//When they are entered it prints the result, and then loops back again and repeats infinitely.
#include <iostream> 
#include <vector> 
#include <cstdlib>

using namespace std; 

// A Binary Tree node 
struct Node 
{ 
	int key; 
	struct Node *left, *right; 
}; 

struct Node * newNode(int k) 
{ 
	cout<<"\nCreating node with value "<<k;
	Node *temp = new Node; 
	temp->key = k; 
	temp->left = temp->right = NULL; 
	return temp; 
} 
struct Node* insert(struct Node* node, int key) 
{ 
    if (node == NULL) return newNode(key); 
  
    if (key < node->key) 
        node->left  = insert(node->left, key); 
    else if (key > node->key) 
        node->right = insert(node->right, key);    
  
    return node; 
} 

struct Node *lca(Node* root, int n1, int n2)  
{  
    if (root == NULL) return NULL;  
  
    
    if (root->key > n1 && root->key > n2)  
        return lca(root->left, n1, n2);  
  
    
    if (root->key < n1 && root->key < n2)  
        return lca(root->right, n1, n2);  
  
    return root;  
}  

int main(int argc, char** argv) 
{ 

	struct Node *root = NULL; 
    root = insert(root,atoi(argv[1])); 

	for (int i =2; i<argc; i++){
		
			int tval = atoi(argv[i]);
			
			insert(root,tval);
	}
	
	int node1=0;
	int node2=0;
	bool finished=false;
	while(!finished){
	cout << "\nEnter 2 nodes you want to calculate lowest common ancestor of (seperated by spaces)\n";
	cin >> node1 >> node2;
	
	cout << "\nLCA("<<node1 <<","<<node2<<") is "<<lca(root, node1, node2)->key; 
}
	return 0; 
} 

File: lowestAncestor/src/lowestAncestor.java
import java.util.*;

class Node {
	int value;
	Node right, left;

	public Node(int val) {
		value = val;
		right = null;
		left = null;
	}

}

public class lowestAncestor {
	Node root;
	public static String toPrint = "";

	public static void main(String[] Args) {
		try {
			lowestAncestor lowest = new lowestAncestor();
			System.out.println("Enter Node values you would like to add (seperated by spaces)");
			Scanner sc = new Scanner(System.in);
			String values = sc.nextLine();
			String[] split = values.split(" ");
			lowest.root = new Node(Integer.valueOf(split[0]));

			for (int i = 1; i < split.length; i++) {

				insert(lowest.root, Integer.valueOf(split[i]));
			}
			System.out.println("Your tree is;");
			System.out.println(prettyPrint(lowest.root, ""));

			System.out.println("");
			System.out.println("Enter the two nodes you want to find the common ancestor of (seperated by spaces)");
			String anNode = sc.nextLine();
			String[] splitter = anNode.split(" ");
			int firstNode = Integer.valueOf(splitter[0]);
			int secondNode = Integer.valueOf(splitter[1]);
			int ancestor = findCommonAncestor(lowest.root, firstNode, secondNode);

			System.out.println("Common ancestor of " + firstNode + " and " + secondNode + " is " + ancestor);

		}

		catch (Exception e) {
			System.out.print("you inputted something incorrectly, please try again");
		}
	}

	public static void insert(Node node, int value) {
		if (value < node.value) {
			if (node.left != null) {
				insert(node.left, value);
			} else {
				node.left = new Node(value);
			}
		} else if (value > node.value) {
			if (node.right != null) {
				insert(node.right, value);
			} else {

				node.right = new Node(value);
			}
		}
	}

	public static String prettyPrint(Node node, String prefix) {

		if (node == null) {
			toPrint = toPrint + prefix + "-null\n";
			return toPrint;
		}
		toPrint = toPrint + prefix + "-" + node.value + "\n";

		prefix = prefix + " |";
		prettyPrint(node.left, prefix);
		StringBuilder sb = new StringBuilder(prefix);
		prefix = prefix.substring(0, prefix.length() - 1);
		prefix = prefix + " ";
		prettyPrint(node.right, prefix);

		return toPrint;
	}

	public static int findCommonAncestor(Node root, int n1, int n2) {

		if (root == null)
			return 0;
		if (root.value > n1 && root.value > n2)
			return findCommonAncestor(root.left, n1, n2);

		if (root.value < n1 && root.value < n2)
			return findCommonAncestor(root.right, n1, n2);

		return root.value;
	}
}
